### ISSUE TITLE
Email address verification

### ISSUE THREAD

**Issue by mzur** [biigle/core#308]
At biigle.de bot sign-ups seem to be increasing. While the existing defenses prevent most basic bots, some get through with seemingly valid names and email addresses and random affiliations (although the names do not fit the email addresses). ~~Consider to implement an option for something like [hCaptcha](https://www.hcaptcha.com) for the sign-up form.~~

Edit: The honeypot/honeytime mechanism seems to be quite effective against bots. This is now mostly about preventing typos in the email addresses.

**Comment by mzur** [comment]
Alternatively we could implement an email address confirmation mechanism. This would also prevent the (increasingly frequent) registrations with an incorrect email address.

**Comment by mzur** [comment]
We should implement email address verification for new users. Make it optional (like the manual confirmation through admins) so it can be turned off. Once an account is verified, the confirmation email is sent to the admins and the account can be manually activated.

BIIGLE could send a montly wrapup of accounts without verified email addresses to the admins, so these accounts can be deleted. Existing accounts should be assumed as verified.

Laravel 9 already inclused an email verification mechanism that we could use.

**Comment by mzur** [comment]
Currently we see lots of students signing up who have weird email naming schemes of their universities. They often make typos and enter incorrect email addresses. I'm raising the priority here because I think the email verification will be more and more important in the future.

**Comment by mzur** [comment]
Lowering priority again as other things are more important right now.

**Comment by mzur** [comment]
Yes, all routes with authentication should be protected. But email verification should be opt-in (with a config option). When a BIIGLE instance enables it, the admin has to run a migration to set all existing users as "verified". All new users go through the new verification, then.

Also, if the BIIGLE instance has `biigle.user_registration_confirmation` enabled, the user has to verify their email first and only when the email is verified, the review/confirmation email is sent to the admins.

Email verification cannot be enabled if `biigle.offline_mode` is enabled.

We don't need a new bot protection mechanism. Our honeypot/honeytime mechanism seems to be quite effective.

**Comment by mzur** [comment]
@dbrembilla I assume you'll pass on this for now? Feel free to unassign so someone else can pick this up.

**Comment by yannik131** [comment]
@mzur 
- I can't find the form for user registration anywhere (biigle.de/register -> localhost:8000/register -> 404), can you tell me how to access it locally?
- Do we want a custom email text or is the [laravel default](https://github.com/laravel/framework/blob/8e641b837db2d4f349b2a663d703fb4f894d3a4a/src/Illuminate/Auth/Notifications/VerifyEmail.php#L62) good enough?
> all routes with authentication should be protected
- So an unverified user trying to log in should see an error toast in the login form requesting the user to verify his email?
- Should unverified users be able to re-request the verification email? Where? At which rate?

**Comment by mzur** [comment]
The registration form is controlled with the `biigle.user_registration` config value.

The default Laravel verification behavior and email templates are enough for now.

We have to distinguish two cases, one with `biigle.user_registration_confirmation` enabled and one with it disabled:

Enabled: A user registers but has the global guest role (i.e. can't create projects, label trees or upload files). The dashboard shows a message that they should verify their email address (similar to the current message that they should wait for admin approval). There they also have a link to resend the verification email. Once they verified the address, they see the old message that they have to wait for admin approval. The admin approval will change their global role. There is no real route protection against unverified users as would typically be the case. The global user role manages authorization.

Disabled: With email verification enabled but user confirmation disabled, new users *also* get the global guest role. They also see the message that they should verify with the link to resend the email. Once they verified, the global role is automatically updated to editor.

This also might make migration easier. Users who already have the global editor role but have `email_verified_at=null` can still keep using BIIGLE as usual. The behavior only changes for new user registrations.

Does this make sense or do you see any holes in this?

Feel free to pick a sensible rate limit for resending verification emails (ideally use the Laravel default).
